runtime/caml/misc.h

```diff
@@ -596,18 +596,19 @@ CAMLextern int caml_snwprintf(wchar_t * buf,
 
 /* Macro used to deactivate address sanitizer on some functions. */
 #define CAMLno_asan
-/* __has_feature is Clang-specific, but GCC defines __SANITIZE_ADDRESS__ and
- * __SANITIZE_THREAD__. */
-#if defined(__has_feature)
+/* __has_feature used to be Clang-specific, and GCC defines
+   __SANITIZE_ADDRESS__ and __SANITIZE_THREAD__ instead.
+   Newer versions of GCC support __has_feature, but do not support all the
+   attributes the Clang does, so we must check for GCC first, because
+   `no_sanitize("thread")` doesn't work on GCC. */
+#if defined(__SANITIZE_ADDRESS__)
+#  undef CAMLno_asan
+#  define CAMLno_asan __attribute__((no_sanitize_address))
+#elif defined(__has_feature)
 #  if __has_feature(address_sanitizer)
```

Comment:
Naive question:

- can we write `if defined(__has_feature) && __has_feature(address_sanitizer)`?
- if yes, do I correctly understand that with that approach the two tests can be written in any order, so we don't need a delicate comment about the ordering that makes this work?

Reply:
I think the problem is that Clang supports `no_sanitize("address")`, but GCC doesn't. GCC only supports `no_sanitize_address` (even though it claims it supports the `address_sanitizer` feature, it doesn't mean it supports all the same ways to turn it off as Clang does.
For Clang `no_sanitize_address` is a deprecated synonym, however AFAICT for GCC that is the only form it supports.

I don't quite understand why, because the manual claims it should support it https://gcc.gnu.org/onlinedocs/gcc-14.1.0/gcc/Common-Function-Attributes.html. Maybe it is a GCC bug?

Reply:
Ah I found the real problem why the suppression was not working in my OCaml 5.2 with GCC.

This commit (which is part of 5.2.0, that part is true) made me believe that `oldify_one` was suppressed: https://github.com/ocaml/ocaml/commit/288ce328d196890b7de247771aaf54dce54d5e01
But later on (still in 5.2!) this other commit removed the suppression (after some bugs in the runtime got fixed):
 https://github.com/ocaml/ocaml/commit/42a633926034d15057b8a6eed7116ab80686273a

Which means the reason why suppressions (seemingly) weren't working with GCC was because the attribute is completely missing on 5.2.
(And then of course if I flip around the order of declarations in the header and do a little test with GCC and look at the disassembly I see that by flipping the order the instrumentation is turned off on my test code, and I assumed that with the default order the instrumentation was on. Although I'm fairly sure I also did a test with the default order and that showed the instrumentation present, but I can't reproduce that now, so I must've been looking at a stale file).

Reply:
Instead this commit is missing on 5.2: https://github.com/ocaml/ocaml/commit/fb6707fa5819e21c9cbab0a96e935a7aea3a84e3

And backporting that commit to 5.2 should fix the TSAN race condition report that I've been seeing.

Sorry for the confusion.